src/visualization/annotation/annotation_base.cpp

```diff
@@ -58,11 +58,32 @@ void AnnotationBase::_sendProgress(double value) {
 
 std::shared_ptr<unity_sframe>
 AnnotationBase::returnAnnotations(bool drop_null) {
-  this->cast_annotations();
-
-  std::shared_ptr<unity_sframe> copy_data =
-      std::static_pointer_cast<unity_sframe>(
+  cast_annotations();
+
+  std::shared_ptr<unity_sframe> copy_data;
+  {
+    // append undefined values
+    if (m_data_na && m_data_na->size()) {
+      auto cast_type = m_data->select_column(m_annotation_column)->dtype();
+      auto na_type = m_data_na->select_column(m_annotation_column)->dtype();
+      if (cast_type != na_type) {
+        auto data_sarray = std::static_pointer_cast<unity_sarray>(
+            m_data_na->select_column(m_annotation_column));
+        auto integer_annotations =
```

Comment:
These aren't necessarily integer annotations, right? Couldn't they be string annotations (depending on cast_type)?

Reply:
Yeah, but the first line `cast` will try to probe whether the annotation can be convertible to integer or not. This is just to align with it.